### DISCUSSION TITLE
Side Effects

### DISCUSSION THREAD

**Discussion by eballeste** [spatie/laravel-model-states#163] [Q&A] [answered]
When reading about custom transition classes to handle side effects there is a note that says the following:

> Note: be careful not to have too many side effects within a transition. If you're injecting many dependencies, it's probably a sign that you should refactor your code and use an event-based system to handle complex side effects.

I have several questions (laravel noob)
- Is this only the case when using dependency injection in transition classes?
- What are the pitfalls of using too many custom transition classes for the side effects? 
- How can you tell when you have too many?
- What are some examples of an event-based system? Is this some type of pattern? Is it meant to replace the state machine pattern or work together with it?

**Comment by brendt** [comment, marked as answer]
> Is this only the case when using dependency injection in transition classes?

It's usually a good indicator, since many dependencies mean many things happen within that class.

> What are the pitfalls of using too many custom transition classes for the side effects?
> How can you tell when you have too many?

I can't say anything to the amount of transaction classes. If your business requires many transactions then you can't prevent that

> What are some examples of an event-based system? Is this some type of pattern? Is it meant to replace the state machine pattern or work together with it?

I was thinking about event sourcing and using aggregate roots as standalone state machines. I can't go into more detail here. I'm actually creating a course on the topic (https://event-sourcing-laravel.com/) and there's lots of great reading material online.
